### ISSUE TITLE
Font Awesome 5

### ISSUE THREAD

**Issue by buschtoens** [martndemus/ember-font-awesome#143]
[Font Awesome 5](https://fontawesome.com/) is almost here and I gotta say: I love it!

The icons look better than ever and being able to chose between **solid**, **regular** and **light** styles is _awesome_. I'm a proud owner of a Font Awesome 5 Pro license and thus already have access to their private repository: [FortAwesome/Font-Awesome-Pro](https://github.com/FortAwesome/Font-Awesome-Pro)

Not trying to advertise here, but you can still get a license for a heavily discounted price of $40, but only for a limited time.

Moving on: I would love to have `ember-font-awesome` be ready for FA5 the day it is released. This issue is meant to gather ideas / wishes and coordinate, if someone wants to help. FA5 comes with a whole load of new features that I would love to support.

- [ ] Distribution type (for lack of a better name)
  - [ ] [**Single SVGs / SVG Framework**](https://github.com/FortAwesome/Font-Awesome-Pro/tree/master/svgs)
  - [ ] [Icon Fonts](https://github.com/FortAwesome/Font-Awesome-Pro/tree/master/webfonts)
  - [ ] [SVG sprites](https://github.com/FortAwesome/Font-Awesome-Pro/tree/master/sprites)
- [ ] Pro Features
  - [ ] How do users provide Pro assets?
  - [ ] CDN
  - [ ] different icon styles (solid, regular, light)
  - [ ] duotone colored icons
- [ ] Accessibility
  - [ ] [`aria-label`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute) attribute: parse labels from [`icons.yml`](https://github.com/FortAwesome/Font-Awesome-Pro/blob/04e79052d4dc4032040026ba262003f753d2ad61/svgs/icons.yml) and use as a default
- [ ] icon subsetting
- [ ] backwards compatibility

As always, any and all help is greatly appreciated. If you have further ideas or wishes please let me know and I'll add them to this list.

**Comment by gossi** [comment]
Cool. Thanks for pushing ember-font-awesome. My first idea would be to get in contact with the guys at FortAwesome about their plans to release FA5 for each framework and strive for a coordinated effort between community and the developers of FA itself (maybe even a new home for this addon ?; ref #136).

Second are the new features FA5 provides (you outlined them already) and at which location they are adressed from an ember point of view (e.g. distribution type is something I'd probably expect to set myself in `ember-cli-build.js`). Others regarding the API itself, which needs to be expanded regarding the new features of FA5, though there are for sure different ways how to achieve them (Regarding markup I prefer the short strings, e.q. `fw` over `fixedWidth` 😉 ).

Pro features: That is something that should possible align with the distribution of FA pro assets, so probably out-of-the-scope for the community addon.

I'm happily looking forward to this.

**Comment by buschtoens** [comment]
> My first idea would be to get in contact with the guys at FortAwesome about their plans to release FA5 for each framework and strive for a coordinated effort between community and the developers of FA itself (maybe even a new home for this addon ?; ref #136).

Good idea! I have just mailed FortAwesome and asked them what their plans are and if they'd like to cooperate with us. I also suggested moving the repository to the FortAwesome organization.

**Comment by robmadole** [comment]
@buschtoens @gossi yep, just got the message in our inbox and replied. This is a great idea and we are interested. We aren't to the point where we are actively developing anything yet (we haven't started on an Ember lib) so this is great timing.

For now I'll volunteer to be a resource on the FA side of things as things come up (I can answer questions but don't have time to contribute code yet).

**Comment by jrjohnson** [comment]
@buschtoens I'm happy to help with this effort. I think it would be awesome to have this addon ready to go on FA5 launch day as (possibly) the first framework to offer full support.

@robmadole is the fontawesome team planning to host the other promised frameworks in your github organization? If so I *do* think we should move this code to match. Otherwise it has the potential to feel abandoned or out of place.

**Comment by robmadole** [comment]
@jrjohnson we haven't 100% decided how we will organize the frameworks. It makes a lot of sense to have them all under the same organization (`FortAwesome`) but that detracts from the ownership that an individual can have if most of the work for the framework is done by them (this project is a bit of an exception it seems since martndemus is wanting to get out from under it)

We (the Font Awesome team) will officially support the frameworks, contribute code, manage issues and releases, document and call them out as officially supported, and in general take a major ownership role. BUT we want community involvement and to share ownership with developers who have specialized knowledge and are willing to do so.

This is all up for discussion, too :D This Ember project is the first to reach out to us and if y'all have any ideas we are listening!

**Comment by martndemus** [comment]
Happy to transfer owner ship, just say yes!

**Comment by robmadole** [comment]
Ok, couple of updates:

1. We've decided that the official libraries are going to be hosted in the `FortAwesome` organization
1. We've accelerated our efforts on ES6/CommonJS/AMD modules and Beta 2 is going to include these to start getting community feedback

So, @buschtoens and @jrjohnson I'm going to add you both as collabs on our private repositories that we are publishing. When either of you get time I would love for you to check them out and give your thoughts. We can proceed from there regarding everything else.

Anybody else who is a collab on this project and wants access to the NPM private repos just let me know.

@martndemus we ain't ready yet for the transfer because we need to think through the name change (we need packages to be called `[tool]-fontawesome` and repos to be `[tool]-fontawesome-js`) and that complicates things.

**Comment by robmadole** [comment]
@buschtoens @jrjohnson you should have access to:

* https://www.npmjs.com/package/@fortawesome/fontawesome
* https://www.npmjs.com/package/@fortawesome/fontawesome-solid
* https://www.npmjs.com/package/@fortawesome/fontawesome-brands
* https://www.npmjs.com/package/@fortawesome/fontawesome-regular
* https://www.npmjs.com/package/@fortawesome/fontawesome-light

**Comment by buschtoens** [comment]
Awesome sauce!

I'm on vacation right now, but I'll check in next week. 

@jrjohnson (and everybody else who would like join in on this) feel free to message me on the Ember Community Slack for quicker chats.

**Comment by jrjohnson** [comment]
I got into this a bit today and here is what I'm seeing so far:

First of all I was able to get a working display with only minor adjustments to the addon. This is really good news for me (though I know @buschtoens at least suspected this was going to be the case).

Necessary Adjustmnets:
- `fonts` directory is now `webfonts`
- need to import both `font-awesome-core.css` and `font-awesome-solid.css` 


I also got the new SVG framework working by pushing into the vendor tree and including:
- `fontawesome.js`
- `packs/fontawesome-solid.js`

Again - good news. @robmadole and the FA team have done some really solid work to make this pretty straightforward.

Now the other news...
I think the biggest issues to tackle is going to be configuration. There are three ways to access FA5:
1) _CSS and webfonts_ (the classic way with CSS and a font file)
2) _SVG through a provided JS framework_ (the preferred method of the FA team and only 22 kb compressed)
3) _SVG Sprites_ (no JS or CSS required but will not work in IE without some kind of polyfil)

Each of these has slightly different use cases and import requirement. Only the svg sprites would require any overhaul to the output of existing components and might be better handled with a `{{fa-sprite}}` component.

Some of the import cases can actually be handled with the existing options like `includeFontAwesomeAssets`, and `includeFontFiles` especially if we add `includeSvgFiles` and `includeSvgFramework`.

Additionally we also need to deal with there being different styles (solid, regular, thin, duo, etc...) for every icon would we want to just allow one style for an application or allow choosing a style in the component `{{fa-icon style='regular'}}`? The second case is going to require even more configuration options to control what assets are actually included though maybe the work in #146 will make that unnecessary?

What I'm getting at with all the text is that there are a lot of configuration possibilities and I'm wondering if we need or even want to actually support them all. Given that ember is a convention driven framework should we just use the SVG JS framework which is the FA team's preferred method, abandon the old font files, and get into sprites only as an additional feature?

**Comment by cibernox** [comment]
For the work I'm doing in this might be interesting. Where can I read more about SVG sprites? I'm working on optimizing the output as much as possible.

I managed to optimize the CSS stripping unused classes (not as much as possible yet, but it's a start). However I've been doing some research on how do the same with the fonts and remove unused glyphs, but it's tricky. I've only found some python wrappers around the FontForge library.

However with SVGs, I have the feeling that it could be much easier.

**Comment by gossi** [comment]
> Additionally we also need to deal with there being different styles (solid, regular, thin, duo, etc...) for every icon would we want to just allow one style for an application or allow choosing a style in the component {{fa-icon style='regular'}}? The second case is going to require even more configuration options to control what assets are actually included though maybe the work in #146 will make that unnecessary?

That might be the first goto solution to this. I, for example might tend to use the thin style everytime, so putting up `style="thin"` would piss me off very soon :wink: Another solution I had, would be to subclass the components `{{fa-thin ...}}` `{{fa-regular ...}}` which basically set the `style` attribute on the parent class (`{{fa-icon}}` as parent) and that's it. Another idea is to set a default in the config. Similarly to the `includeXyzWhatever` config variables, there could also be `defaultStyle` which is regular by default but could be overwritten. In the end, all my ideas here are combinable, thought should be thought through for better refinement.

**Comment by jrjohnson** [comment]
@cibernox unfortunately everything for FA5 is currently private for kickstarter backers.  @robmadole would it be ok if I shared the beta5 downloads with select contributors here?

@gossi I agree with you - in our app, for consistency, we're only probably only going to use one icon style across the board specifying that in the options would seem to make a lot of sense. I also like the `{{fa-STYLE}}` option. We could (by default) include the assets for only the solid style `{{fa-icon}}` and then put some fences inside `{{fa-thin}}` and output a warning if the assets were not present with some instructions for enabling it.

**Comment by gossi** [comment]
> We could (by default) include the assets for only the solid style {{fa-icon}} and then put some fences inside {{fa-thin}} and output a warning if the assets were not present with some instructions for enabling it.

That might correct under a wrong assumption. For the majority of icons you would be right using one style but you might want to use the one or the other icon in a different style which seems absolutely legit to me.

**Comment by buschtoens** [comment]
I would include all styles by default and let the user overwrite that.

`{{fa-$style}}` as a wrapper for `{{fa-icon style=$style}}` is a neat idea.

**Comment by robmadole** [comment]
@jrjohnson yeah, share the beta releases with select collaborators, that's fine by me.

**Comment by jrjohnson** [comment]
Thanks @robmadole.
I asked for (and we got) an #e-font-awesome channel in ember slack. If you need these files please join and we can talk about this and more transient details of this update there.

**Comment by jrjohnson** [comment]
Long comment forthcoming. Sorry in advance.

After looking at the new AST transforms and the way FA5 is provided I propose we do this:

 1) Include all packs in the build by default and provide `{{fa-$style}}` and `{{fa-icon style=$style}}` in many cases we can simple remove those icons which are not used. We can also provide options to not include certain packs.
2) Drop support CSS, SCSS, LESS and fonts and not include support for the new sprite format. Thus choosing a single convention over the current configuration options.
3) Handle transforming component syntax into HTML ourselves.

**Reasoning**
While FA5 provides 3 different ways to present icons the default and most compatible way is to inline the SVG. FA5 is providing a JS library to convert `<i class='fa fa-home'></i>` into 
```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 576 512"><path fill="currentColor" d="M488 312.7V456c0 13.3-10.7 24-24 24H348c-6.6 0-12-5.4-12-12V356c0-6.6-5.4-12-12-12h-72c-6.6 0-12 5.4-12 12v112c0 6.6-5.4 12-12 12H112c-13.3 0-24-10.7-24-24V312.7c0-3.6 1.6-7 4.4-9.3l188-154.8c4.4-3.6 10.8-3.6 15.3 0l188 154.8c2.7 2.3 4.3 5.7 4.3 9.3zm83.6-60.9L488 182.9V44.4c0-6.6-5.4-12-12-12h-56c-6.6 0-12 5.4-12 12V117l-89.5-73.7c-17.7-14.6-43.3-14.6-61 0L4.4 251.8c-5.1 4.2-5.8 11.8-1.6 16.9l25.5 31c4.2 5.1 11.8 5.8 16.9 1.6l235.2-193.7c4.4-3.6 10.8-3.6 15.3 0l235.2 193.7c5.1 4.2 12.7 3.5 16.9-1.6l25.5-31c4.2-5.2 3.4-12.7-1.7-16.9z"></path></svg>
```

While this JS transform from the FA team is great, we can actually skip that step entirely and just do the transform from `{{fa-icon 'home'}}` straight to the SVG and skip `<i>` entirely as FA5 provides the icon paths like:

```
var faHome = { prefix: 'fas', iconName: 'home', icon: [576, 512, [], "f015", "M488 312.7V456c0 13.3-10.7 24-24 24H348c-6.6 0-12-5.4-12-12V356c0-6.6-5.4-12-12-12h-72c-6.6 0-12 5.4-12 12v112c0 6.6-5.4 12-12 12H112c-13.3 0-24-10.7-24-24V312.7c0-3.6 1.6-7 4.4-9.3l188-154.8c4.4-3.6 10.8-3.6 15.3 0l188 154.8c2.7 2.3 4.3 5.7 4.3 9.3zm83.6-60.9L488 182.9V44.4c0-6.6-5.4-12-12-12h-56c-6.6 0-12 5.4-12 12V117l-89.5-73.7c-17.7-14.6-43.3-14.6-61 0L4.4 251.8c-5.1 4.2-5.8 11.8-1.6 16.9l25.5 31c4.2 5.1 11.8 5.8 16.9 1.6l235.2-193.7c4.4-3.6 10.8-3.6 15.3 0l235.2 193.7c5.1 4.2 12.7 3.5 16.9-1.6l25.5-31c4.2-5.2 3.4-12.7-1.7-16.9z"] };
```

This has the additional benefits of allowing us to present a nice error message when the icon name is wrong, will be performant since there is no need to lookup or load any additional assets, and will avoid the dreaded empty squares when the font doesn't load. 

Just as it is in 4.0.0 if there are no dynamic icons used in the app we can toss all the definitions we don't need so the end result will just be these SVG blocks in static HTML that ember doesn't need to parse.

The only downside I see to the SVG approach could be a case where the same icon was included many times in an app. In that case there would be a duplication of the SVG block since it wasn't being used. I think that is an OK tradeoff at this point and when browser support for loading from a sprite sheet improves we can switch to that option fairly seamlessly.

Once @cibernox finishes the current AST transform work (and with a final `4.0.0` release) I can start this work so we are ready to ship a `5.0-beta` once FA5 development goes public and publishes NPM packages that everyone can access.

**Comment by robmadole** [comment]
@jrjohnson thanks for the detailed look into this!

First up: npm modules are released!

[@fortawesome/fontawesome](https://www.npmjs.com/package/@fortawesome/fontawesome)
[@fortawesome/fontawesome-free-solid](https://www.npmjs.com/package/@fortawesome/fontawesome-free-solid)
[@fortawesome/fontawesome-free-brands](https://www.npmjs.com/package/@fortawesome/fontawesome-free-brands)

When you say "Include all packs in the build" are you talking about just the Free solid and brands packs? (The regular and light are for paying customers only)

You mentioned:

> While this JS transform from the FA team is great, we can actually skip that step entirely

I'm wondering if you've seen [this](https://www.npmjs.com/package/@fortawesome/fontawesome#fontawesomeiconicondefinition-params)?

This is the rough documentation we have for the npm module `@fortawesome/fontawesome`.

This will build the HTML, a node, or an abstract JSON structure and would save a lot of work and keep this project up-to-date with the main library. The Vue and React components are both leaning on this lib right now and they became pretty trivial to implement by leveraging `fontawesome.icon(faPlus).abstract`.

> Just as it is in 4.0.0 if there are no dynamic icons used in the app we can toss all the definitions we don't need

Perfect. We also took this approach with our core lib. However, we made allowances for situations where you needed dynamic icons with [this feature](https://gist.github.com/robmadole/708be3e5e19d4ffebae90c076739cd86#fontawesomelibraryaddicondefinitions)

> so the end result will just be these SVG blocks in static HTML that ember doesn't need to parse.

You can use `fontawesome.icon(faPlus).html` in this case.

> The only downside I see to the SVG approach could be a case where the same icon was included many times in an app. In that case there would be a duplication of the SVG block since it wasn't being used. I think that is an OK tradeoff at this point and when browser support for loading from a sprite sheet improves we can switch to that option fairly seamlessly.

We will eventually build SVG sprite support into the JS framework to handle this duplicate icon situation. The performance benefits of sprites are pretty fantastic in some browsers and use cases. (Think a page with thousands of table rows all having an edit icon and no efficient paging mechanism).

**Comment by jrjohnson** [comment]
@robmadole sweet! `fontawesome.icon(faPlus).html` looks great. 

I didn't realize that regular and light were paid only. How are you going to deliver those icons to customers? Are they going to be in a different NPM repo? Is there going to be a license key or something?

We will only include the free stuff by default, but I would like to understand how we can allow pro customers to use those icons easily in this addon.

**Comment by robmadole** [comment]
@jrjohnson our current plan is to have a private NPM repository. We will most likely have revocable access keys for authentication.

**Comment by knownasilya** [comment]
What is the progress here so far? Just wondering, since I am looking at using FA5 and start using this addon at the same time.

**Comment by jrjohnson** [comment]
We need to ship a final v4 of this addon first. I believe that #189 is the last obstacle to getting that done. 
Once that is shipped I believe v5 to support FA5 with inline SVG icons will follow fairly quickly since most of the hard work for re-writing in place was done already by @cibernox in v4.

**Comment by robmadole** [comment]
FA5 has been released now and I would like to update everyone on our current thoughts regarding Ember and Font Awesome (and our official support for a component).

We've decided that we are going to make some reference implementations that use `@fortawesome/fontawesome` as the base package and support the SVG version of Font Awesome only. We've completed two now: [Vue](https://www.npmjs.com/package/@fortawesome/vue-fontawesome) and [React](https://www.npmjs.com/package/@fortawesome/react-fontawesome).

And we are continuing with the Ember version next. @mlwilkerson is going to be the lead on this. We are going to start from scratch / green field on this as opposed to usurping `ember-font-awesome`.

Our implementation will get slotted into https://fontawesome.com/get-started/advanced-options. Once the `ember-font-awesome` project is FA5-ready we will add it to that page and to our deeper docs. I'd be happy to dive into this decision if anyone would like. (We also are probably going to do some branding/badges that indicated a project is "officially recommended by Font Awesome" or similar)

OK, update done. I also wanted to get @mlwilkerson looped into this so that he can collaborate. @jrjohnson and @buschtoens do you mind if we throw ideas at you and the rest of this group as we work through things? Also, let us know if we can support you all in any way as you add support for FA5.

**Comment by mlwilkerson** [comment]
Hey everybody. I'm catching up on your conversation here. Looking forward to collaboration.

**Comment by gossi** [comment]
thanks @robmadole for the update. My question here is what does this mean for this addon? Since this is like the go-to addon with FA and gained some popularity, will it be moved to `fortawesome` and be maintained by you and will the name stay?

Seen from an ember perspective, how will we consume the FA addon in the future, free and pro version?

**Comment by jrjohnson** [comment]
@robmadole / @mlwilkerson I'm not sure you want to start from scratch here as there is a *lot* of really good work done to handle the performant replacement of components and support for advanced cases. Plus the ember community is very organized with resources like https://emberobserver.com/ to avoid duplication.  I know we already talked about moving the repo and development to your github org, we could also consider a clean start in `master` here, but other than the README all of the code here would end up being included in your addon as well.

Maybe we can talk about pros / cons in slack? We can join you workspace if you have one, or feel free to jump in our, instructions here https://github.com/martndemus/ember-font-awesome#sos-looking-for-help

**Comment by mlwilkerson** [comment]
@jrjohnson I've joined your Slack channel.

**Comment by buschtoens** [comment]
@robmadole / @mlwilkerson 

Thanks for being so open and collaborative and also congrats on launching officially!

If I understood correctly, you want to build an official Ember reference implementation from a clean slate first and then, as soon as `ember-font-awesome` hits v5 compatibility, add us as an endorsed alternative?

To be honest, I think that sounds like unnecessary extra work. @cibernox has built some *really nifty* things into this addon, especially looking forward to the SVG framework. Starting over and making users chose between two addons would surely cause confusion.

I also think releasing a v5 compatible alpha / beta of this addon is right around the corner. Today we landed #189, which solved a compatibility issue and was the last blocking step to an official v4 release, of which I'm prepping a RC right now.

**Comment by donaldwasserman** [comment]
I want to follow up here with some perspective of a long-time listener, first time caller. 

We use this addon on 5 different actively maintained projects across ~15 developers. 

If the "greenfield" approach @robmadole / @mlwilkerson are talking about maintains the existing `{{fa-icon}}` api, and it's a matter of just a matter of updating package.json or maybe some configuration, great! If not, it's going to be pretty problematic to make those updates across various applications + shared component libraries (forms, buttons, etc)

I hope whatever gets settled focused on maintaining the existing api, with an easy upgrade path for organizations like mine who are happy where the cost of rewriting our apps is far higher than the cost of purchasing licenses.

**Comment by robmadole** [comment]
Oh, shoot. Sorry folks. We've asked @mlwilkerson if he's willing to be a liaison. We'll rethink this (I honestly didn't expect this response). Mike can share our goals and concerns and let's see how it shakes out. We're going to put a hold on any further development on our end. Reaching a consensus with this project is now priority numeral uno. Thanks, everyone for being so...gentle... 💯

**Comment by jamesdixon** [comment]
Hi everyone -- wanted to see if there was any movement on this? Cheers!

**Comment by mlwilkerson** [comment]
Hi @jamesdixon, we've resumed the discussion in the EmberJS Community Slack team, #e-font-awesome channel. After some discussion and prototyping I've proposed a way forward, but we're awaiting further consideration from key stakeholders.

**Comment by jrjohnson** [comment]
Since slack has eaten some of our discussion and other stuff was private I just wanted to summarize where we are at with this right now.

I've been unable to figure out how to use the AST transform work to build the `<svg>` elements for FA5. It requires changing both the attributes of the `<svg>` element as well as the contents of its children (to add a `<path>`). In light of this we're currently looking at a green field FA5 addon that will use standard components and can be released relatively soon then we can work to bring the AST and tree shaking code here into that at a later date.

If anyone else has time to take a shot at brining FA5 into `master` here please get in touch over slack and I will be happy to walk you through the FA javascript API or help with any info you might need.

**Comment by mlwilkerson** [comment]
I've just pushed a pre-release of `@fortawesome/ember-fontawesome` with official support from the Font Awesome team for Font Awesome 5.

`ember install @fortawesome/ember-fontawesome`

See the repo for a README (this is still prerelease, so we'd love help with bringing this to maturity):
https://github.com/FortAwesome/ember-fontawesome

Thanks @jrjohnson for guidance.

We're glad to incorporate the developments from `ember-font-awesome` that fit with the FA5 paradigm.

Note that this release does include tree shaking to slim down those big icon packs, but does not (yet) include those build-time AST transform optimizations from `ember-font-awesome`.

**Comment by mlwilkerson** [comment]
A little context for that pre-release and the repo: it's a milestone in the plan that I've been working, toward these goals:

1. Shipping an official component under the `@fortawesome` npm scope.

1. Retaining as far as possible the template API already established in `ember-font-awesome`, 
i.e. `{{fa-icon ...}}`

1. Incorporate and carry forward any of the build-time optimizations in `ember-font-awesome` that the community wants to see move forward into the FA5 world. (Given that we have other ways to do sub-setting/tree-shaking in the FA5 world, it's not yet clear to us how advantageous a build-time SVG-rendering approach would be, and for which use cases, but we're open to all of it—just seeking community help to PR the appropriate bits into place.)

It's not currently a goal of ours (the Font Awesome team) to support FA4 with the `@fortawesome/ember-fontawesome` addon.

**Comment by NullVoxPopuli** [comment]
Font Awesome 5 is pretty cool. It'd be great to have it supported :-D
